crates/bevy_reflect/src/set.rs

```diff
@@ -0,0 +1,557 @@
+    pub fn value_type_id(&self) -> TypeId {
+        self.value_type_id
+    }
+
+    /// Check if the given type matches the value type.
+    pub fn value_is<T: Any>(&self) -> bool {
+        TypeId::of::<T>() == self.value_type_id
+    }
+
+    /// The docstring of this set, if any.
+    #[cfg(feature = "documentation")]
+    pub fn docs(&self) -> Option<&'static str> {
+        self.docs
+    }
+}
+
+/// An ordered set of reflected values.
+#[derive(Default)]
+pub struct DynamicSet {
+    represented_type: Option<&'static TypeInfo>,
+    values: Vec<Box<dyn Reflect>>,
+    indices: HashMap<u64, usize>,
+}
+
+impl DynamicSet {
+    /// Sets the [type] to be represented by this `DynamicSet`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given [type] is not a [`TypeInfo::Set`].
+    ///
+    /// [type]: TypeInfo
+    pub fn set_represented_type(&mut self, represented_type: Option<&'static TypeInfo>) {
+        if let Some(represented_type) = represented_type {
+            assert!(
+                matches!(represented_type, TypeInfo::Set(_)),
+                "expected TypeInfo::Set but received: {:?}",
+                represented_type
+            );
+        }
+
+        self.represented_type = represented_type;
+    }
+
+    /// Inserts a typed value into the set.
+    pub fn insert<V: Reflect>(&mut self, value: V) {
+        self.insert_boxed(Box::new(value));
+    }
+}
+
+impl Set for DynamicSet {
+    fn get(&self, value: &dyn Reflect) -> Option<&dyn Reflect> {
+        self.indices
+            .get(&value.reflect_hash().expect(hash_error!(value)))
+            .map(|index| &**self.values.get(*index).unwrap())
+    }
+
+    fn get_mut(&mut self, value: &dyn Reflect) -> Option<&mut dyn Reflect> {
+        self.indices
+            .get(&value.reflect_hash().expect(hash_error!(value)))
+            .cloned()
+            .map(|index| &mut **self.values.get_mut(index).unwrap())
+    }
+
+    fn get_at(&self, index: usize) -> Option<&dyn Reflect> {
+        self.values.get(index).map(|value| &**value)
+    }
+
+    fn len(&self) -> usize {
+        self.values.len()
+    }
+
+    fn iter(&self) -> SetIter {
+        SetIter::new(self)
+    }
+
+    fn drain(self: Box<Self>) -> Vec<Box<dyn Reflect>> {
+        self.values
+    }
+
+    fn clone_dynamic(&self) -> DynamicSet {
+        DynamicSet {
+            represented_type: self.represented_type,
+            values: self
+                .values
+                .iter()
+                .map(|value| value.clone_value())
+                .collect(),
+            indices: self.indices.clone(),
+        }
+    }
+
+    fn insert_boxed(&mut self, mut value: Box<dyn Reflect>) -> bool {
+        match self
+            .indices
+            .entry(value.reflect_hash().expect(hash_error!(value)))
+        {
+            Entry::Occupied(entry) => {
+                let old_value = self.values.get_mut(*entry.get()).unwrap();
+                std::mem::swap(old_value, &mut value);
```

Comment:
Isn't this just `self.values[*entry.get()] = value;`?

Reply:
Yeah, thanks! Nice catch. I guess this is a leftover from a impl where we were returning the replaced value